File: 567_perm_in_string.ts
const getPos = (s: string) => {
    return s.charCodeAt(0) - "a".charCodeAt(0);
};

const stringHash = (s: string) => {
    const hashMap = new Array(26).fill(0);
    for (let i = 0; i < s.length; i++) {
        hashMap[getPos(s[i])]++;
    }
    return hashMap;
};

function checkInclusion(s1: string, s2: string): boolean {
    const hash1 = stringHash(s1);

    let low = 0,
        high = 0;

    const hash2 = new Array(26).fill(0);

    for (high = 0; high < s2.length; high++) {
        hash2[getPos(s2[high])]++;
        while (low < high && hash1[getPos(s2[low])] < hash2[getPos(s2[low])]) {
            hash2[getPos(s2[low])]--;
            low++;
        }
        if (hash2.toString() === hash1.toString()) {
            return true;
        }
    }

    return false;
}

File: 567_perm_in_string.py
class Solution:

    def get_pos(self, char: str):
        return ord(char) - ord('a')

    def hash_string(self, s: str):
        arr = [0] * 26
        for char in s:
            arr[self.get_pos(char)] += 1
        return arr

    def checkInclusion(self, s1: str, s2: str) -> bool:
        hash1 = self.hash_string(s1)
        low = 0
        high = 0
        hash2 = [0] * 26
        for high in range(len(s2)):
            hash2[self.get_pos(s2[high])] += 1

            while (low < high and hash2[self.get_pos(s2[low])] > hash1[self.get_pos(s2[low])]):
                hash2[ord(s2[low]) - ord('a')] -= 1
                low += 1

            if (hash1 == hash2):
                return True
        return False
